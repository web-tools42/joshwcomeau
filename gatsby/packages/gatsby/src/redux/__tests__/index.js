const _ = require(`lodash`)

const writeToCache = jest.spyOn(require(`../persist`), `writeToCache`)
const { saveState, store, readState } = require(`../index`)

const {
  actions: { createPage },
} = require(`../actions`)

const mockWrittenContent = new Map()
jest.mock(`fs-extra`, () => {
  return {
    writeFileSync: jest.fn((file, content) =>
      mockWrittenContent.set(file, content)
    ),
    readFileSync: jest.fn(file => mockWrittenContent.get(file)),
  }
})

describe(`redux db`, () => {
  const initialComponentsState = _.cloneDeep(store.getState().components)

  beforeEach(() => {
    store.dispatch(
      createPage(
        {
          path: `/my-sweet-new-page/`,
          // seems like jest serializer doesn't play nice with Maps on Windows
          component: `/Users/username/dev/site/src/templates/my-sweet-new-page.js`,
          // The context is passed as props to the component as well
          // as into the component's GraphQL query.
          context: {
            id: `123456`,
          },
        },
        { name: `default-site-plugin` }
      )
    )

    writeToCache.mockClear()
    mockWrittenContent.clear()
  })

  it(`expect components state to be empty initially`, () => {
    expect(initialComponentsState).toEqual(new Map())
  })

  it(`should write cache to disk`, async () => {
    await saveState()

    expect(writeToCache).toBeCalled()

    // reset state in memory
    store.dispatch({
      type: `DELETE_CACHE`,
    })
    // make sure store in memory is empty
    expect(store.getState().components).toEqual(initialComponentsState)

    // read data that was previously cached
    const data = readState()

    // make sure data was read and is not the same as our clean redux state
    expect(data.components).not.toEqual(initialComponentsState)

    // yuck - loki and redux will have different shape of redux state (nodes and nodesByType)
    expect(_.omit(data, [`nodes`, `nodesByType`])).toMatchSnapshot()
  })
})
