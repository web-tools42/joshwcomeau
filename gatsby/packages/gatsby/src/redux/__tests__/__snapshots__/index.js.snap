// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redux db should write cache to disk 1`] = `
Object {
  "componentDataDependencies": Object {
    "connections": Map {},
    "nodes": Map {},
  },
  "components": Map {
    "/Users/username/dev/site/src/templates/my-sweet-new-page.js" => Object {
      "componentPath": "/Users/username/dev/site/src/templates/my-sweet-new-page.js",
      "isInBootstrap": true,
      "pages": Set {
        "/my-sweet-new-page/",
      },
      "query": "",
    },
  },
  "pageDataStats": Map {},
  "staticQueryComponents": Map {},
  "status": Object {
    "plugins": Object {},
  },
  "webpackCompilationHash": "",
}
`;
